### DISCUSSION TITLE
Azure Artifacts NPM Configuration

### DISCUSSION THREAD

**Discussion by munkyjunky** [renovatebot/renovate#8608] [Request Help] [answered]
I'm trying to configure renovate against a series of git repos in Azure. Each repo uses a private NPM registry within Azure artifacts, and while I can get the onboarding PR started I'm struggling with getting renovate access to our private NPM registry. 

- I'm running my own renovate bot from the docker image. 
- each repository has it's own `.npmrc` that points to the registry location
- each developer has their own `.npmrc` in their home directory with credentials to access the registry
- checking in credentials to the repository `.npmrc` isn't an option, so we can't combine the two into one
- encoding the credential `.npmrc` and placing in config seems to causes issues because it's too long

My understanding was that by mounting my own `npmrc` with credentials to `/home/ubuntu/.npmrc` in the container, renovate would pick this up and use the credentials - however, I can't seem to get it to work. Currently, all I'm getting are 401 errors against the registry.

Project npmrc:
```
registry=https://my-registry.pkgs.visualstudio.com/_packaging/foobar/npm/registry/
always-auth=true
```

Personal npmrc (which I mapped to `/home/ubuntu/.npmrc`):
```
; begin auth token
//my-registry.pkgs.visualstudio.com/_packaging/foobar/npm/registry/:username=someUsername
//my-registry.pkgs.visualstudio.com/_packaging/foobar/npm/registry/:_password=[BASE64 ENCODED ACCESS TOKEN]
//my-registry.pkgs.visualstudio.com/_packaging/foobar/npm/registry/:email=somevaluetostopnpmcomplaining
//my-registry.pkgs.visualstudio.com/_packaging/foobar/npm/:username=someUsername
//my-registry.pkgs.visualstudio.com/_packaging/foobar/npm/:_password=[BASE64 ENCODED ACCESS TOKEN]
//my-registry.pkgs.visualstudio.com/_packaging/foobar/npm/:email=somevaluetostopnpmcomplaining
; end auth token
```

Any guidance on the correct configuration on this would be appreciated. I'm probably just missing something simple, hopefully.

**Comment by viceice** [comment, marked as answer]
checkout this: https://docs.renovatebot.com/setup-azure-devops/

**Reply by munkyjunky** [reply]
Doesn't seem to be changing anything - have also tried with the hostname set as this:
```
  {
      "hostName": "MY-ORG.pkgs.visualstudio.com",
      "username": "MY-ORG", // as the username is in npmrc
      "token": "foobar" // tried both plain and base64 token
    }
```

**Reply by japj** [reply]
I connected to an on-premise azure DevOps 2020 server yesterday, so perhaps this might help you:

The way I did it was to use some information form https://docs.renovatebot.com/private-modules/ and to actually add code to config.js that reads the .npmrc file we use during the regular build and add it to the module exports section. That way the renovatebot should be able to use it.

**Reply by munkyjunky** [reply]
Managed to get it working with the following config:

```
const TOKEN = process.env.AZURE_TOKEN;

module.exports = {
  "platform": "azure",
  "token": TOKEN,
  "endpoint": "https://ORG.visualstudio.com/",
  "repositories": ["team/my-repo"],
  "hostRules": [
    {
      "hostName": "ORG.pkgs.visualstudio.com",
      "username": "",
      "password": TOKEN
    }
  ]
}
```

**Reply by viceice** [reply]
this should be added to the docs section 🙃
